### ISSUE TITLE
Inconsistent behavior on undefined object property reference

### ISSUE THREAD

**Issue by wjordan** [mde/ejs#35]
Minimal test:
```js
require('ejs').render('<%- data.test %>', {data:{}})
```
In ejs <=1.0.0, this returns `''` (empty string).
In ejs 2.x, this returns `'undefined'`.

This breaks some existing templates, so unless there's an intentional reason to make this breaking change (and it's clearly documented), I would like to see consistency with the <=1.0.0 behavior.

**Comment by TimothyGu** [comment]
(BTW you don't need the `eval()`)

Will fix.

**Comment by TimothyGu** [comment]
Technically this is a breaking change. @mde what do you think?

To fix this, I would need to make EJS output templates like one of these:

1. EJS 1 style:

   ```js
var buf = []
[...]
;buf.push(data.test)
[...]
;return buf.join('')
```

2. Ternary operator:

   ```js
var __output = ''
[...]
;__output += (data.test) == null ? '' : data.test
[...]
;return __output
```

I'll do some benchmark on these.

**Comment by TimothyGu** [comment]
```
ejs v1 comp x 24,390 ops/sec ±1.21% (95 runs sampled)
ejs v2 comp orig (__output) x 7,175 ops/sec ±0.57% (97 runs sampled)
ejs v2 comp buf x 7,111 ops/sec ±0.33% (101 runs sampled)
ejs v2 comp __output x 6,927 ops/sec ±0.39% (103 runs sampled)

ejs v1 x 344,302 ops/sec ±0.55% (100 runs sampled)
ejs v1 !_with x 1,560,845 ops/sec ±0.49% (96 runs sampled)
ejs v2 orig x 143,544 ops/sec ±0.52% (100 runs sampled)
ejs v2 orig !_with x 1,453,962 ops/sec ±0.67% (97 runs sampled)
ejs v2 buf x 192,506 ops/sec ±0.97% (100 runs sampled)
ejs v2 buf !_with x 1,328,785 ops/sec ±1.62% (92 runs sampled)
ejs v2 out x 131,761 ops/sec ±1.15% (94 runs sampled)
ejs v2 out !_with x 1,457,213 ops/sec ±3.15% (88 runs sampled)
```

```js
  <% for (var i = 0; i < data.length; i++) { %>
    <% var objects = data[i] %>
    <h1><%= data[i].title %></h1>
    <% for (var j = 0; j < objects.length; j++) { %>
      <% var obj = objects[j] %>
      <% if (obj.passed) { %>
        <p>Passed: <%= obj.name %></p>
        <p>Value: <%= obj.value %></p>
      <% } else { %>
        <p>Failed: <%= obj.name %></p>
        <p>Value: <%= obj.value %></p>
      <% } %>
    <% } %>
  <% } %>
```

```js
    { data: [
      {
        title: "Test title",
        objects: [
          {
            name: 'this',
            value: 21032,
            passed: false
          }, {
            name: 'test',
            value: 0,
            passed: true
          }
        ]
      }, {
        title: "Another data",
        objects: [
          {
            name: 'that',
            value: 1012021,
            passed: true
          }, {
            name: 'tset',
            value: -2,
            passed: false
          }
        ]
      }
    ]}
```

According to the principle of the least slow I think *if* we are going to make this change it should use `var buf = []` (the original EJS <=1 approach).

**Comment by TimothyGu** [comment]
Now I read more and more EJS 1 code, I really think it's fairly reusable as EJS v2. The OO stuff and expensive regex matching is really dragging the compilation speed down. EJS v1 depends purely on `indexOf` which, believe it or not, is *fast*.

**Comment by mde** [comment]
I seem to remember fixing this for escaped output a long time ago in my original codebase. We can do some rewriting for perf, but it's an easy fix for unescaped output too. I'll push a fix, and we can talk about rewrite and refactor in another GH issue or PR.

**Comment by mde** [comment]
Fixed in master in a09d026a41f44b7f34c2e8cf17b86589611d4d51.
